File: tools/token_bench/programs/04_concurrency/concurrency.go
package main

import (
	"fmt"
	"sync"
)

func fib(n int) int {
	if n < 2 {
		return n
	}
	return fib(n-1) + fib(n-2)
}

func main() {
	results := make([]int, 4)
	var wg sync.WaitGroup
	for i := 0; i < 4; i++ {
		wg.Add(1)
		go func(idx int) {
			defer wg.Done()
			results[idx] = fib(27)
		}(i)
	}
	wg.Wait()
	r := 0
	for _, v := range results {
		r += v
	}
	fmt.Printf("parallel: %d\n", r)

	results2 := make([]int, 4)
	var wg2 sync.WaitGroup
	for i := 0; i < 4; i++ {
		wg2.Add(1)
		go func(idx int) {
			defer wg2.Done()
			results2[idx] = fib(27)
		}(i)
	}
	wg2.Wait()
	total := 0
	for _, v := range results2 {
		total += v
	}
	fmt.Printf("pmap: %d\n", total)

	ch := make(chan int, 10)
	go func() {
		for i := 0; i < 10; i++ {
			ch <- i * i
		}
	}()
	for i := 0; i < 10; i++ {
		fmt.Println(<-ch)
	}
}

File: tools/token_bench/programs/04_concurrency/concurrency.js
const { Worker, isMainThread, parentPort, workerData } = require("worker_threads");

function fib(n) {
    if (n < 2) return n;
    return fib(n - 1) + fib(n - 2);
}

if (!isMainThread) {
    parentPort.postMessage(fib(workerData));
} else {
    async function runParallel() {
        const promises = Array.from({length: 4}, () =>
            new Promise((resolve, reject) => {
                const w = new Worker(__filename, {workerData: 27});
                w.on("message", resolve);
                w.on("error", reject);
            })
        );
        const results = await Promise.all(promises);
        const r = results.reduce((a, b) => a + b, 0);
        console.log(`parallel: ${r}`);

        const promises2 = [27, 27, 27, 27].map(n =>
            new Promise((resolve, reject) => {
                const w = new Worker(__filename, {workerData: n});
                w.on("message", resolve);
                w.on("error", reject);
            })
        );
        const results2 = await Promise.all(promises2);
        const total = results2.reduce((a, b) => a + b, 0);
        console.log(`pmap: ${total}`);

        const {Channel} = require("worker_threads");
        const values = [];
        for (let i = 0; i < 10; i++) values.push(i * i);
        for (const v of values) console.log(v);
    }
    runParallel();
}
